Explicitly export symbols via __all__

### What feature should we add?

Currently when importing symbols from package scope causes an error with mypy when running in strict mode:

```python
from web3.providers import WebsocketProviderV2
# error: Module "web3.providers" does not explicitly export attribute "WebsocketProviderV2"  [attr-defined]
```

It would be nice if the package `__init__.py` files could include an __all__ attribute to silence this error by explicitly marking the symbols as being re-exported in the new namespace.

--- a/newsfragments/3410.feature.rst
+++ b/newsfragments/3410.feature.rst
@@ -1,0 +1,1 @@
+Provide explicit ``__all__`` exports for providers in `web3/providers/__init__.py`; update `web3/__init__.py` to include all provider classes including base classes.

--- a/web3/__init__.py
+++ b/web3/__init__.py
@@ -1,6 +1,15 @@
 from eth_account import Account  # noqa: E402
 import sys
 
+from web3.providers import (
+    AsyncBaseProvider,
+    AutoProvider,
+    BaseProvider,
+    JSONBaseProvider,
+    PersistentConnectionProvider,
+)
+
+
 if sys.version_info.major == 3 and sys.version_info.minor < 8:
     import pkg_resources
 
@@ -19,6 +28,7 @@ from web3.providers.async_rpc import (  # noqa: E402
     AsyncHTTPProvider,
 )
 from web3.providers.eth_tester import (  # noqa: E402
+    AsyncEthereumTesterProvider,
     EthereumTesterProvider,
 )
 from web3.providers.ipc import (  # noqa: E402
@@ -35,13 +45,21 @@ from web3.providers.websocket import (  # noqa: E402
 
 __all__ = [
     "__version__",
+    "Account",
+    # web3:
     "AsyncWeb3",
     "Web3",
+    # providers:
+    "AsyncBaseProvider",
+    "AsyncEthereumTesterProvider",
+    "AsyncHTTPProvider",
+    "AutoProvider",
+    "BaseProvider",
+    "EthereumTesterProvider",
     "HTTPProvider",
     "IPCProvider",
+    "JSONBaseProvider",
+    "PersistentConnectionProvider",
     "WebsocketProvider",
     "WebsocketProviderV2",
-    "EthereumTesterProvider",
-    "Account",
-    "AsyncHTTPProvider",
 ]

--- a/web3/providers/__init__.py
+++ b/web3/providers/__init__.py
@@ -8,6 +8,10 @@ from .base import (
     BaseProvider,
     JSONBaseProvider,
 )
+from .eth_tester import (
+    AsyncEthereumTesterProvider,
+    EthereumTesterProvider,
+)
 from .ipc import (
     IPCProvider,
 )
@@ -24,3 +28,18 @@ from .persistent import (
 from .auto import (
     AutoProvider,
 )
+
+__all__ = [
+    "AsyncBaseProvider",
+    "AsyncEthereumTesterProvider",
+    "AsyncHTTPProvider",
+    "AutoProvider",
+    "BaseProvider",
+    "EthereumTesterProvider",
+    "HTTPProvider",
+    "IPCProvider",
+    "JSONBaseProvider",
+    "PersistentConnectionProvider",
+    "WebsocketProvider",
+    "WebsocketProviderV2",
+]

